src/search/operator_counting/delete_relaxation_constraints.h

```diff
@@ -0,0 +1,73 @@
+#ifndef OPERATOR_COUNTING_DELETE_RELAXATION_CONSTRAINTS_H
+#define OPERATOR_COUNTING_DELETE_RELAXATION_CONSTRAINTS_H
+
+#include  "constraint_generator.h"
+
+#include "../task_proxy.h"
+
+#include <memory>
+
+namespace lp {
+class LPVariable;
+}
+
+namespace options {
+class Options;
+}
+
+namespace operator_counting {
+class DeleteRelaxationConstraints : public ConstraintGenerator {
+    bool use_time_vars;
+    bool use_integer_vars;
+
+    /* [U_o] Is op part of the relaxed plan?
+       Binary, indexed with op.id */
+    std::vector<int> lp_var_id_op_used;
+
+    /* [R_f] Is fact <V,v> reached by the relaxed plan?
+       Binary, indexed with var.id, value */
+    std::vector<std::vector<int>> lp_var_id_fact_reached;
+
+    /* [F_{o,f}] Is o the first achiever of fact <V,v> in the relaxed plan?
+       Binary, indexed with op.id, var.id, value */
+    std::vector<std::vector<std::vector<int>>> lp_var_id_first_achiever;
+
+    /* [T_o] At what time is o used first?
+       {0, ..., |O|}, indexed with op.id */
+    std::vector<int> lp_var_id_op_time;
+
+    /* [T_f] At what time is <V,v> first achieved?
+       {0, ..., |O|}, indexed with var.id, value */
+    std::vector<std::vector<int>> lp_var_id_fact_time;
+
+    /* Indices of constraints that change in every state
+       Indexed with var.id, value */
+    std::vector<std::vector<int>> constraint_ids;
+
+    /* The state that is currently used for setting the bounds.
+       Remembering this makes it faster to unset the bounds when the state
+       changes. */
+    std::vector<FactProxy> last_state;
+
+    int get_var_op_used(OperatorProxy op);
+    int get_var_fact_reached(FactProxy f);
+    int get_var_first_achiever(OperatorProxy op, FactProxy f);
+    int get_var_op_time(OperatorProxy op);
+    int get_var_fact_time(FactProxy f);
+    int get_constraint_id(FactProxy f);
+
+    void create_auxiliary_variables(TaskProxy task_proxy,
+        named_vector::NamedVector<lp::LPVariable> &variables);
```

Comment:
I think a `using LPVariables = named_vector::NamedVector<lp::LPVariable>;` typedef would make the code nice to read (same for constraints), but feel free to ignore.

Reply:
There are some complications with this: `LPVariable` is forward-declared in the header but `LPConstraint` is not. I now added the forward declaration for constraints just to keep the parallel structure, even if the header doesn't require the typedef. It would also be nice to have this typedef more globally (e.g., next to the definition of `LinearProgram`) but as far as I know, typedefs cannot have forward declarations. Please have a look at the new version and let me know what you think.

Reply:
Can't we have the typedefs directly above the declaration of LinearProgram?

Reply:
I would prefer that as well, but the issue is that you cannot forward declare typedefs. So here in this header file, we would have to include `../lp/lp_solver.h` instead of forward-declaring the type and including it in the cc file as we do now.

Reply:
In other places we avoided this problem by having a header with only types (`pdbs/types.h`) but this might be overkill here.

Reply:
Ah, now I see the problem. Then let's leave things as they are in the patch now.